### DISCUSSION TITLE
How to invalidate first provider of chained providers from last provider?

### DISCUSSION THREAD

**Discussion by jeiea** [rrousselGit/riverpod#2651] [Q&A] [answered]
Consider there are provider A, B, C.

A is FutureProvider fetching data from remote, and B is stateless provider that depends on A, and C is stateless provider that depends on B.

Consumer depends on C only. If I want to invalidate A for recreating C, should Consumer invalidate A directly? I'm not sure if that's desirable, as it makes Consumer know A, and changes in B, C result changes in Consumer.

**Comment by csbence10** [comment]
I think yes, you should invalidate provider A. This way all of the providers depending on A will rebuild. If you just invalidated provider C, basically nothing would happen, because the "underlying data" is the same.

**Reply by jeiea** [reply]
I agree, but it gets tricky when B starts relying on other providers like A1, A2, and A3. Everywhere I use C, I will write code that invalidates A1, A2, and A3, and I don't think that's predictable.

**Comment by rrousselGit** [comment, marked as answer]
Refactor `C` to a class-variant, and expose a custom `invalidate` method on your Notifier. THen have C.invalidate invalidate the internal provider of your choice.
